src/core/tsi/s2a/handshaker/s2a_tsi_handshaker.cc

```diff
@@ -16,34 +16,327 @@
+  // TODO(mattstev): do I need to add nullptr checks?
+  gpr_free(result->in_traffic_secret);
+  gpr_free(result->out_traffic_secret);
+  gpr_free(result->spiffe_id);
+  gpr_free(result->hostname);
+  gpr_free(result->unused_bytes);
+  gpr_free(result);
+}
+
+static const tsi_handshaker_result_vtable s2a_result_vtable = {
+    s2a_handshaker_result_extract_peer,
+    s2a_handshaker_result_create_zero_copy_grpc_protector, nullptr,
+    s2a_handshaker_result_get_unused_bytes, s2a_handshaker_result_destroy};
+
+// TODO(mattstev): I don't use the application protocol and local identity
+// fields from |response|, or the attributes field from the peer identity.
 tsi_result s2a_tsi_handshaker_result_create(s2a_SessionResp* response,
                                             bool is_client,
-                                            tsi_handshaker_result** result) {
-  return TSI_UNIMPLEMENTED;
+                                            tsi_handshaker_result** self) {
+  if (self == nullptr || response == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultNullptrArguments);
+    return TSI_INVALID_ARGUMENT;
+  }
+  const s2a_SessionResult* handshake_result = s2a_SessionResp_result(response);
+  if (handshake_result == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultEmpty);
+    return TSI_FAILED_PRECONDITION;
+  }
+  const s2a_Identity* peer_identity =
+      s2a_SessionResult_peer_identity(handshake_result);
+  if (peer_identity == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultInvalidPeerIdentity);
+    return TSI_FAILED_PRECONDITION;
+  }
+  const s2a_SessionState* handshake_state =
+      s2a_SessionResult_state(handshake_result);
+  if (handshake_state == nullptr) {
+    gpr_log(GPR_ERROR, kS2ATsiHandshakerResultInvalidSessionState);
+    return TSI_FAILED_PRECONDITION;
+  }
+  upb_strview in_traffic_secret = s2a_SessionState_in_key(handshake_state);
+  upb_strview out_traffic_secret = s2a_SessionState_out_key(handshake_state);
+  if (in_traffic_secret.size != out_traffic_secret.size) {
+    gpr_log(GPR_ERROR, kS2ATrafficSecretSizeMismatch);
+    return TSI_FAILED_PRECONDITION;
+  }
+
+  /** Instantiate S2A TSI handshaker result. **/
+  s2a_tsi_handshaker_result* tsi_result =
+      static_cast<s2a_tsi_handshaker_result*>(
+          gpr_malloc(sizeof(s2a_tsi_handshaker_result)));
+  tsi_result->in_traffic_secret = nullptr;
+  tsi_result->out_traffic_secret = nullptr;
+  tsi_result->spiffe_id = nullptr;
+  tsi_result->hostname = nullptr;
+  tsi_result->unused_bytes = nullptr;
+  tsi_result->unused_bytes_size = 0;
+  tsi_result->is_client = is_client;
+  tsi_result->base.vtable = &s2a_result_vtable;
+  *self = &(tsi_result->base);
+
+  /** Populate fields of |tsi_result| using |peer_identity|. **/
+  if (s2a_Identity_has_spiffe_id(peer_identity)) {
+    upb_strview spiffe_id = s2a_Identity_spiffe_id(peer_identity);
+    tsi_result->spiffe_id =
+        static_cast<char*>(gpr_malloc(spiffe_id.size * sizeof(char)));
```

Comment:
Please use `gpr_zalloc` whenever possible.

Reply:
Changed all `gpr_malloc` instances to `gpr_zalloc` here.